### PR TITLE
Add C++ support to torch.utils.benchmark.Timer

### PR DESCRIPTION
**Pull request by robieta**
This PR makes pretty heavy use of `torch.utils.cpp_extension` to JIT compile snippets. The rest was fairly simple; Timer already defers to an underlying `_timer_cls` for its times, and `collect_callgrind` just needs something that it can subprocess execute with appropriate `CALLGRIND_TOGGLE_COLLECT`'s. I didn't initially intend to add the wall time portion, but it just kind of fell out naturally.

Key changes:
1) Expose `CXX_FLAGS` so the JIT'd code can match them. (Namely `-O*` and `-g`)
2) Add an option for `cpp_extension.load` to emit a standalone binary. Currently linux only: @malfet I wasn't sure what the windows analog to `-rpath` was.
3) Bring along the Valgrind headers as part of the wheel, as they're needed for the C++ measure loop.
4) Add `cpp_jit.py` and a couple templates. I tried to group as much of compilation bits as possible, though it has the curious effect of putting the `timeit` C++ under `valgrind_wrapper/`. I organized the intermediate files to be friendly to `cpp_extension`'s caching scheme while still not littering random artifacts, and repeated runs are reasonably snappy.
5) Add a `language=` argument to Timer that accepts `Python` or `C++`, and plumb accordingly.
6) Tie number of displayed lines to `PRINT_OPTS.threshold`. (This is just a papercut that's been bothering me for a while.)
7) Change compat bindings to not touch torch at all, which got rid of the strange behavior that compat bindings increased the instruction counts due to the inclusion of certain headers.

And now for the question that is doubtless on everyone's mind: Yes, the C++ instruction counts appear to be absolutely deterministic. (Yay!)

I've also added an example showing how instruction counts correlate to wall time, and how one could show the "Python tax" by diff-ing equivalent statements in Python and C++.

### CODE REVIEW

**Comment by robieta** [comment]
Build failures are due to my treatment of callgrind headers. (symlink, but then copy during build) Shouldn't be too bad to sort out.

**Comment by ezyang** [comment]
> Bring along the Valgrind headers as part of the wheel, as they're needed for the C++ measure loop.

Haven't read the diff yet. I hope these headers don't conflict with hypothetical system headers someone else might have of Valgrind when compiling with TorchScript.

**Comment by ezyang** [comment]
For future reference, nearly each of the individual changes could have been submitted as a separate diff in a diff stack. That would make the review easier; now I have to go hunting for, e.g., which specific lines are "Tie number of displayed lines to PRINT_OPTS.threshold"

File: torch/utils/benchmark/utils/valgrind_wrapper/callgrind.h [hunk 1]
```diff
@@ -0,0 +1 @@
+/data/users/taylorrobie/repos/pytorch/third_party/valgrind-headers/callgrind.h
```
Comment (ezyang) [inline torch/utils/benchmark/utils/valgrind_wrapper/callgrind.h:1]: fixme

File: torch/utils/benchmark/utils/valgrind_wrapper/valgrind.h [hunk 2]
```diff
@@ -0,0 +1 @@
+/data/users/taylorrobie/repos/pytorch/third_party/valgrind-headers/valgrind.h
```
Comment (ezyang) [inline torch/utils/benchmark/utils/valgrind_wrapper/valgrind.h:1]: fixme

BTW, checked in symlinks don't work correctly on Windows anyway, so you don't want to check these in at all

**Comment by ezyang** [comment]
Adding @peterjc123 re Windows support. It shouldn't be difficult to do.

File: torch/utils/cpp_extension.py [hunk 3]
```diff
@@ -1675,7 +1713,7 @@ def object_file_path(source_file: str) -> str:
     elif IS_WINDOWS:
         ldflags = _nt_quote_args(ldflags)
 
-    ext = 'pyd' if IS_WINDOWS else 'so'
+    ext = 'pyd' if IS_WINDOWS else ('o' if is_standalone else 'so')
```
Comment (ezyang) [inline torch/utils/cpp_extension.py:1716]: The Windows branch here is wrong. pyd = shared library that can be loaded by Python on Windows. You almost definitely want .o here.

File: torch/utils/cpp_extension.py [hunk 4]
```diff
@@ -1555,7 +1584,11 @@ def _run_ninja_build(build_directory: str, verbose: bool, error_prefix: str) ->
         raise RuntimeError(message) from e
 
 
-def _import_module_from_library(module_name, path, is_python_module):
+def _import_module_from_library(module_name, path, is_python_module, is_standalone):
+    if is_standalone:
+        assert not IS_WINDOWS
```
Comment (ezyang) [inline torch/utils/cpp_extension.py:1589]: seems better to raise a NotImplementedError, unless the intention here is that the precondition for calling this function is that you're not on Windows (distinguish between "Developer error" and "we just haven't implemented this yet")

File: torch/utils/cpp_extension.py [hunk 5]
```diff
@@ -1356,7 +1379,12 @@ def _prepare_ldflags(extra_ldflags, with_cuda, verbose):
         if with_cuda:
             extra_ldflags.append('-ltorch_hip' if IS_HIP_EXTENSION else '-ltorch_cuda')
         extra_ldflags.append('-ltorch')
-        extra_ldflags.append('-ltorch_python')
+
+        if is_standalone:
+            extra_ldflags.append(f"-Wl,-rpath,{lib_path}")
```
Comment (ezyang) [inline torch/utils/cpp_extension.py:1384]: FYI, another way to solve this problem is to just set `LD_LIBRARY_PATH` before invoking these executables. rpath is almost certainly more user friendly as it lets people run the binaries by hand.

File: torch/utils/benchmark/examples/callgrind.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
+"""Demonstrate Callgrind collection in both C++ and Python."""
```
Comment (ezyang) [inline torch/utils/benchmark/examples/callgrind.py:1]: Is this file exercised by CI?

File: torch/utils/benchmark/utils/timer.py [hunk 7]
```diff
@@ -21,6 +22,57 @@ def timer() -> float:
     timer = timeit.default_timer
 
 
+PYTHON = "Python"
+CPP = "C++"
```
Comment (ezyang) [inline torch/utils/benchmark/utils/timer.py:26]: Insisting that these be capitalized is not very PyTorch-y (where we say "cuda" not "CUDA")

File: torch/utils/benchmark/examples/callgrind.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        setup="x = torch.ones((1,))",
+    )
+
+    cpp_timer = benchmark_utils.Timer(
+        stmt="auto y = torch::add(x, torch::ones({1}));",
+        setup="torch::Tensor x = torch::ones({1});",
+        language="C++",
+    )
+
+    python_times = python_timer.blocked_autorange(min_run_time=5)
```
Comment (ezyang) [inline torch/utils/benchmark/examples/callgrind.py:19]: This is not really part of this PR, but I really can't tell what `blocked_autorange` means just looking at the method name.

File: torch/utils/benchmark/utils/timer.py [hunk 9]
```diff
@@ -21,6 +22,57 @@ def timer() -> float:
...
+        globals: Dict[str, Any],
+    ) -> None:
+        if timer is not timeit.default_timer:
+            # FIXME: We should be able to automatically detect GPU operations.
+            raise ValueError(
+                "When timing C++, timer must be `timeit.default_timer`. If you "
+                "did not specify `timer=...`, this means that a timer with "
+                "CUDA syncronization was chosen, and you should pass "
+                "`timer=timeit.default_timer` to the Timer's constructor to "
+                "declare that CUDA syncronization is not needed."
```
Comment (ezyang) [inline torch/utils/benchmark/utils/timer.py:44]: *synchronization (misspelled in two places)

But why would C++ imply that you don't need CUDA synchronization? If I have a C++ binary that does GPU operations, I still need to sync to get accurate numbers.

File: torch/utils/benchmark/utils/timer.py [hunk 10]
```diff
@@ -21,6 +22,57 @@ def timer() -> float:
...
+            )
+
+        if len(globals) > 1 or "torch" not in globals:
+            # Timer will add `torch` as a convenience.
+            raise ValueError("C++ timing does not support globals.")
```
Comment (ezyang) [inline torch/utils/benchmark/utils/timer.py:49]: If I somehow explicitly pass an empty globals, won't this still raise a value error? That doesn't seem right.

Comment (ezyang) [reply torch/utils/benchmark/utils/timer.py:49]: The whole concept of globals here doesn't make sense (the torch global doesn't make sense either, the corresponding C++ concept doesn't really have any relation to what you're doing here.) Seems better to just not have globals at all in this case.

File: torch/utils/benchmark/utils/timer.py [hunk 11]
```diff
@@ -21,6 +22,57 @@ def timer() -> float:
...
+                "Swapping out `torch` is not allowed for C++ snippets."
+            )
+
+        self._stmt: str = textwrap.dedent(stmt)
+        self._setup: str = textwrap.dedent(setup)
+        self._timeit_module: Optional[cpp_jit.TimeitModuleType] = None
+
+    def timeit(self, number: int) -> float:
+        if self._timeit_module is None:
+            self._timeit_module = cast(
```
Comment (ezyang) [inline torch/utils/benchmark/utils/timer.py:65]: Hmm, how come you need a cast here...

File: torch/utils/benchmark/utils/timer.py [hunk 12]
```diff
@@ -135,10 +189,25 @@ def __init__(
...
 
+        self._language = language
+        if language not in (PYTHON, CPP):
+            raise ValueError(f"Invalid language `{language}`. (Choices: {PYTHON}, {CPP})")
+
+        if language == CPP:
+            assert self._timer_cls is timeit.Timer, "_timer_cls has already been swapped."
+            self._timer_cls = cast(
+                # CPPTimer does not inherit from timeit.Timer, however it implements
+                # `timeit` which is sufficient for our purposes.
```
Comment (ezyang) [inline torch/utils/benchmark/utils/timer.py:205]: Do you want a Protocol here? https://mypy.readthedocs.io/en/stable/protocols.html

File: torch/utils/cpp_extension.py (hunk 3 above)
Comment (peterjc123) [reply torch/utils/cpp_extension.py:1716]: Some common extension translations between Windows and Linux
```
pyd/dll = so
obj = o
lib = a
```

File: torch/utils/cpp_extension.py [hunk 13]
```diff
@@ -1714,6 +1754,8 @@ def _write_ninja_file(path,
     `library_target`: Name of the output library. Can be None; in that case,
                       we do no linking.
     `with_cuda`: If we should be compiling with CUDA.
+    `is_standalone`: If we should produce a standalone binary rather than a
```
Comment (peterjc123) [inline torch/utils/cpp_extension.py:1757]: Actually, I don't quite understand the argument `is_standalone` here. What does `binary` in the description mean? An executable? Or an object file? If you refer to the latter one here, then how are you gonna use it? Does Python allow users to load the functions from an object file?

File: torch/utils/cpp_extension.py (hunk 3 above)
Comment (peterjc123) [reply torch/utils/cpp_extension.py:1716]: By the way, the file extension of an executable is `.exe` in case if you need that.

File: torch/utils/cpp_extension.py (hunk 13 above)
Comment (peterjc123) [reply torch/utils/cpp_extension.py:1757]: If you refer to the former one, then the solution is simple. Just change the extension to `.exe`. And I think it will work out of box.

File: test/test_utils.py [hunk 14]
```diff
@@ -642,5 +645,57 @@ def test_assert_true(self):
...
+                os.path.join(build_dir, f"standalone_load_test.{ext}")
+            )
+
+            for shell in [True, False]:
+                r = subprocess.run(
+                    [exec_path],
+                    shell=shell,
+                    stdout=subprocess.PIPE,
+                )
+                self.assertEqual(r.returncode, 0)
```
Comment (peterjc123) [inline test/test_utils.py:685]: The current Windows exit code `3221225781 = 0xc0000135 = STATUS_DLL_NOT_FOUND`. 
To workground this, You 'll need to patch the `PATH` variable. Take a look at https://github.com/pytorch/pytorch/blob/master/torch/__init__.py#L46-L77 and https://github.com/pytorch/pytorch/blob/master/torch/__init__.py#L122.
```
if IS_WINDOWS:
    run_env = os.environ.copy()
    run_env["PATH"] = additional_paths + run_env["PATH"]

...
subprocess.run(...., env=run_env)
```

Comment (robieta) [reply test/test_utils.py:685]: I think the question is what the API should be. It's moot for my use since Valgrind is Linux only, but someone might find it standalone C++ JIT useful for other applications. Maybe return two strings: `PATH additions` and `exec path`, where the former is just empty on Linux? Or disallow altogether on Windows? (That's what I initially did, but I figured I'd use CI to make sure that it wouldn't work out of the box.) I don't have strong feelings one way or another.

File: torch/utils/cpp_extension.py (hunk 5 above)
Comment (peterjc123) [reply torch/utils/cpp_extension.py:1382]: If we use `LD_LIBRARY_PATH` here, then the API is unified on both Linux and Windows. But alternatively, if we use rpath on Linux, as far as I know, there isn't anything equivalent for the MSVC linker. But we can build a bootstrap program that sets the corresponding PATH. What do you think, @ezyang?

File: torch/utils/benchmark/utils/valgrind_wrapper/callgrind.h (hunk 1 above)
Comment (robieta) [reply torch/utils/benchmark/utils/valgrind_wrapper/callgrind.h]: Done.

File: torch/utils/benchmark/utils/valgrind_wrapper/valgrind.h (hunk 2 above)
Comment (robieta) [reply torch/utils/benchmark/utils/valgrind_wrapper/valgrind.h]: Done. (I misread what was going on in `setup.py`; the symlinks are now gone.)

File: torch/utils/cpp_extension.py (hunk 4 above)
Comment (robieta) [reply torch/utils/cpp_extension.py]: In this case it was an internal assert. (Earlier code should have prevented that bit from being reached, this just checked that it did.) I put those asserts anywhere that would stop making sense if the top `NotImplementedError` was removed.

File: torch/utils/benchmark/examples/callgrind.py (hunk 6 above)
Comment (robieta) [reply torch/utils/benchmark/examples/callgrind.py:1]: It isn't. The benchmark utils examples are in a transitional state, serving as pseudo-documentation while the real documentation is being created.

File: torch/utils/benchmark/utils/timer.py (hunk 7 above)
Comment (robieta) [reply torch/utils/benchmark/utils/timer.py:31]: Done.

File: torch/utils/benchmark/examples/callgrind.py (hunk 8 above)
Comment (robieta) [reply torch/utils/benchmark/examples/callgrind.py:19]: Are you familiar with the `timeit.Timer.autorange` method?

File: torch/utils/benchmark/utils/timer.py (hunk 9 above)
Comment (robieta) [reply torch/utils/benchmark/utils/timer.py:63]: Sorry. I'm really bad about misspelling that.

It's doing the opposite; it's saying that it doesn't support CUDA yet. (There doesn't seem to be an analog to `torch.cuda. synchronize` in the C++ API.) Long term we should just be able to detect CUDA operations during the warmup and choose the timer appropriately, but that's for a future PR.

File: torch/utils/benchmark/utils/timer.py (hunk 10 above)
Comment (robieta) [reply torch/utils/benchmark/utils/timer.py:67]: This was working around the fact that there was a convenience `globals.set_default("torch", torch)`. A cleaner solution is to just separate Python specific and C++ specific logic in `__init__` and simplify downstream checks, which I've done.

File: torch/utils/benchmark/utils/timer.py (hunk 11 above)
Comment (robieta) [reply torch/utils/benchmark/utils/timer.py:75]: The type checker can't deduce the type. I've added a comment explaining why.

File: torch/utils/benchmark/utils/timer.py (hunk 12 above)
Comment (robieta) [reply torch/utils/benchmark/utils/timer.py]: Yes. That is **exactly** what I wanted.

**Comment by robieta** [comment]
> For future reference, nearly each of the individual changes could have been submitted as a separate diff in a diff stack. That would make the review easier; now I have to go hunting for, e.g., which specific lines are "Tie number of displayed lines to PRINT_OPTS.threshold"

Fair point. I don't like rewriting history which unfortunately rules out `ghstack` (right?), but I should definitely get in the habit of stacking branches. Grouping too many things in one PR is a bad habit; thanks for calling me out on it.

**Comment by ezyang** [comment]
> Fair point. I don't like rewriting history which unfortunately rules out ghstack (right?), but I should definitely get in the habit of stacking branches. 

Yes, one of the tricky things about doing stacked development is when you discover you need to retro old commits. You can avoid these situations by being super sure about your early diffs, and landing them ASAP. The good thing about a small stack is people are more likely to approve your small early commits, so you can start landing them early.

File: torch/utils/benchmark/utils/timer.py (hunk 9 above)
Comment (ezyang) [reply torch/utils/benchmark/utils/timer.py:63]: (1) There should be a `torch::cuda::synchronize`; the functions in `torch::cuda` have always been missing bits and pieces here and there and could use some love. (2) It's really easy to do synchronize in C++. Here's the implementation of the Python binding of `_cuda_synchronize`:

```
PyObject * THCPModule_cudaSynchronize(PyObject *_unused, PyObject *noargs)
{
  HANDLE_TH_ERRORS
  c10::cuda::device_synchronize();
  Py_RETURN_NONE;
  END_HANDLE_TH_ERRORS
}

```

File: torch/utils/cpp_extension.py (hunk 5 above)
Comment (ezyang) [reply torch/utils/cpp_extension.py:1391]: Making an entire bootstrap program seems like overcall, but setting PATH when we invoke this stuff seems pretty reasonable.

File: setup.py [hunk 15]
```diff
@@ -325,8 +326,16 @@ def check_file(f):
 
     # Use copies instead of symbolic files.
     # Windows has very poor support for them.
-    sym_files = ['tools/shared/_utils_internal.py']
-    orig_files = ['torch/_utils_internal.py']
+    sym_files = [
+        'tools/shared/_utils_internal.py',
+        'torch/utils/benchmark/utils/valgrind_wrapper/callgrind.h',
+        'torch/utils/benchmark/utils/valgrind_wrapper/valgrind.h',
```
Comment (ezyang) [inline setup.py:332]: Are you sure these files will get packaged correctly when we make Python wheels?

File: torch/utils/benchmark/utils/timer.py [hunk 16]
```diff
@@ -135,16 +201,32 @@ def __init__(
...
         if not isinstance(stmt, str):
             raise ValueError("Currently only a `str` stmt is supported.")
 
-        # We copy `globals` to prevent mutations from leaking, (for instance,
-        # `eval` adds the `__builtins__` key) and include `torch` if not
-        # specified as a convenience feature.
         globals = dict(globals or {})
-        globals.setdefault("torch", torch)
-        self._globals = globals
+        if language.lower() == PYTHON.lower():
```
Comment (ezyang) [inline torch/utils/benchmark/utils/timer.py:210]: This wasn't how I expected you were going to fix it. This is code is especially strange to me because PyTorch *doesn't* accept the capitalized variants of device names:

```
>>> torch.zeros(0, device='CPU')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Expected one of cpu, cuda, mkldnn, opengl, opencl, ideep, hip, msnpu, xla device type at start of device string: CPU
```

File: torch/utils/benchmark/utils/timer.py [hunk 17]
```diff
@@ -21,6 +27,64 @@ def timer() -> float:
     timer = timeit.default_timer
 
 
+PYTHON = "Python"
+CPP = "C++"
+
+
+class InnerTimer(Protocol):
```
Comment (ezyang) [inline torch/utils/benchmark/utils/timer.py:34]: Would be nice to have an explanation what this is, and what types are known to implement this Protocol.

**Comment by ezyang** [comment]
> Fair point. I don't like rewriting history which unfortunately rules out ghstack (right?), but I should definitely get in the habit of stacking branches. Grouping too many things in one PR is a bad habit; thanks for calling me out on it.

For example, why is this PR review taking a long time? Because I have to juggle review for each orthogonal change on the PR at the same time. This makes me grumpy, it makes me slow, and it makes me more likely to miss things on the first round of review, which means you don't even get all the relevant feedback first time around. Review time isn't really linear in PR size, it's more like quadratic or something.

File: torch/utils/benchmark/utils/timer.py (hunk 11 above)
Comment (ezyang) [reply torch/utils/benchmark/utils/timer.py:75]: A better strategy, when performance doesn't matter (which I don't think it does here), is instead do `assert isinstance(r, cpp_jit.TimeitModuleType)`. This has the added benefit of actually calling you out at runtime when you get it wrong. I much prefer Haskell's naming for its equivalent function: `unsafeCoerce`

File: setup.py (hunk 15 above)
Comment (ezyang) [reply setup.py:332]: In particular, I don't see any update to `package_data`

File: torch/utils/benchmark/utils/timer.py [hunk 18]
```diff
@@ -2,11 +2,17 @@
 
 import timeit
 import textwrap
-from typing import Any, Callable, Dict, List, NoReturn, Optional
+from typing import cast, Any, Callable, Dict, List, NoReturn, Optional, Type, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from typing import Protocol
+else:
+    Protocol = object
```
Comment (ezyang) [inline torch/utils/benchmark/utils/timer.py:10]: Euh, is this a BC thing?

File: torch/utils/benchmark/utils/timer.py (hunk 16 above)
Comment (ezyang) [reply torch/utils/benchmark/utils/timer.py:210]: Also, if you're going to go through the trouble of converting an internal rep, might as well define an actual honest to goodness Enum for it; it's more type safe.

File: torch/utils/cpp_extension.py (hunk 5 above)
Comment (robieta) [reply torch/utils/cpp_extension.py:1391]: So what is the verdict on what `load` should return? (And should we keep `rpath` in Linux, or require users to use `LD_LIBRARY_PATH` for parity?)

File: torch/utils/benchmark/utils/timer.py (hunk 9 above)
Comment (robieta) [reply torch/utils/benchmark/utils/timer.py:63]: My concern was replicating `_lazy_init()`, which is not so simple.

File: setup.py (hunk 15 above)
Comment (robieta) [reply setup.py:332]: They aren't. I mistakenly took the entry out when I removed the symlinks; I'll put it back.

File: torch/utils/benchmark/utils/timer.py (hunk 18 above)
Comment (robieta) [reply torch/utils/benchmark/utils/timer.py:10]: MyPy has a fit if you do:
```
try:
    from typing import Protocol
except ImportError:
    from typing_extensions import Protocol
```

and the `Protocol = object` trick is how other parts of the code handled that. I quite like your suggestion above of using an assert rather than a cast, although it would (I think) require the rather comical:
```
if TYPE_CHECKING:
    from typing import Protocol
else:
    try:
        from typing import Protocol
    except ImportError:
        from typing_extensions import Protocol
```

File: torch/utils/benchmark/utils/valgrind_wrapper/timer_interface.py [hunk 19]
```diff
@@ -568,20 +577,36 @@ def run(args: List[str], **kwargs: Any) -> Tuple[CompletedProcessType, str]:
...
-            with open(script_file, "wt") as f:
-                f.write(self._construct_script(
-                    task_spec,
-                    globals=GlobalsBridge(globals, data_dir),
-                    number=number,
-                    error_log=error_log,
-                    stat_log=stat_log,
-                    bindings=self._bindings_module))
+                run_loop_cmd = [
+                    cast(str, run_loop_exec),
```
Comment (ezyang) [inline torch/utils/benchmark/utils/valgrind_wrapper/timer_interface.py:605]: the casts here also look dodgy af

File: torch/utils/benchmark/utils/valgrind_wrapper/cpp_jit.py [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+    from typing import Protocol
+else:
+    Protocol = object
+
+import torch
+from torch.utils import cpp_extension
+
+
+SOURCE_ROOT = os.path.split(os.path.abspath(__file__))[0]
+with open(os.path.join(SOURCE_ROOT, "timer_timeit_template.cpp"), "rt") as f:
```
Comment (ezyang) [inline torch/utils/benchmark/utils/valgrind_wrapper/cpp_jit.py:26]: This seems incredibly unlikely to work in fbcode

File: torch/utils/benchmark/utils/valgrind_wrapper/cpp_jit.py [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+
+with open(os.path.join(SOURCE_ROOT, "timer_callgrind_template.cpp"), "rt") as f:
+    CXX_CALLGRIND_TEMPLATE = f.read()
+
+
+BUILD_ROOT = os.path.join(
+    torch._appdirs.user_cache_dir(appname="benchmark_utils_jit"),
+    f"build_{uuid.uuid4()}".replace("-", "")
+)
+os.makedirs(BUILD_ROOT, exist_ok=True)
```
Comment (ezyang) [inline torch/utils/benchmark/utils/valgrind_wrapper/cpp_jit.py:38]: doing file system writes on module import: BIG NO NO (file system reads are not very good either)

File: torch/utils/benchmark/utils/valgrind_wrapper/cpp_jit.py [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+WRITE_SOURCE_LOCK = threading.Lock()
+atexit.register(shutil.rmtree, BUILD_ROOT)
```
Comment (ezyang) [inline torch/utils/benchmark/utils/valgrind_wrapper/cpp_jit.py:40]: If these really are temporary files, it seems much better to actually put them in the system temporary directory, so they get cleared out even if your process segfaults or something.

File: torch/utils/benchmark/utils/valgrind_wrapper/cpp_jit.py [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+        build_directory=build_dir,
+        extra_cflags=CXX_FLAGS,
+        extra_include_paths=extra_include_paths,
+        is_python_module=not is_standalone,
+        is_standalone=is_standalone,
+    )
+
+    return (
+        # This is purely for MyPy and human comprehension.
+        cast(str, output) if is_standalone else
```
Comment (ezyang) [inline torch/utils/benchmark/utils/valgrind_wrapper/cpp_jit.py:128]: You can do better than this, using Mypy overloads and Literal. Define two overloads of this function and have it return a different type depending on if `is_standalone: Literal[True]` and False.

File: torch/utils/benchmark/utils/valgrind_wrapper/cpp_jit.py [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+if conda_prefix is not None:
+    extra_include_paths = [os.path.join(conda_prefix, "include")]
+extra_include_paths.append(SOURCE_ROOT)
+
+
+# MyPy has no way of knowing the methods on a C module, so we must
+# define API classes which mirror the PyBind11 methods.
+class TimeitModuleType(Protocol):
+    def timeit(self, number: int) -> float:
+        ...
```
Comment (ezyang) [inline torch/utils/benchmark/utils/valgrind_wrapper/cpp_jit.py:60]: The more conventional way to do this is to define a `pyi` file for the C bindings. You can see examples of this in `torch/_C`. If you're doing the pybind11 by literally grabbing this module and then stuffing extra bindings on it... well, don't do that, it's super confusing for code readers who are like, "wait, where the heck did this binding come from"

File: torch/utils/benchmark/utils/valgrind_wrapper/cpp_jit.py (hunk 21 above)
Comment (robieta) [reply torch/utils/benchmark/utils/valgrind_wrapper/cpp_jit.py:38]: Gah! Thanks for catching this.

**Comment by robieta** [comment]
Closing as this has proved too big a PR to review efficiently. I will incorporate feedback and re-open as multiple PRs.
